docs/src/pages/ThemingSpecification.mdx

```diff
@@ -0,0 +1,301 @@
+| Adding new design term in the parent requires re-composing all slots components                                   | NO                                          | We can just override styles in the slot's styles inside the parents                                                                            | YES                                      | We need to re-compose if we want new design terms in the slots components                                               |
+| Requires knowledge of the internals of another component, which creates coupling                                  | PARTIALLY                                   | The parent can define styles only for the root (no other slots), but it needs to know some details for the root's styles of the slot component | NO                                       | All styles are defined in the composed component                                                                        |
+| Easy to swap a slot into another component which might want all its own styles.                                   | NO                                          |                                                                                                                                                | YES                                      |
+
+##### Variants
+
+Variants are just mechanism for allowing users do add or remove some design terms from their components.
+
+In order to enable this, the compose mechanism should provide to the users adding/removing some of the design terms. This can be easily done, by allowing the user to define mapping from props to design terms when they define the composed components.
+
+```
+export interface FluentButton extends ButtonProps {
+  primary?: boolean
+  circular?: boolean
+}
+
+// these code snippet is just illustrating a possible
+// mechanism of how the design tokens can be created
+// based on the props
+const FluentButton = compose(Button, {
+  displayName: 'FluentButton',
+  mapPropsToTerms: props => ({
+    primary: props.primary,
+    circular: props.circular,
+    hasIcon: !!props.icon,
+  }),
+})
+```
+
+##### Animations
+
+Animation should be defined as part of the theme, but they should not be all loaded if they are not used. Child themes should be able to customize them.
+
+```
+// TBD
+```
+
+##### Icons
+
+Icons should be defined as part of the theme, but they should not loaded if not used. Child themes should be able to customize them.
+
+```
+// TBD
+```
+
+##### Files and packaging
+
+```
+// themes/fluent/globalTokens.ts
+
+const globalTokens = {
+  colors: {
+    brandForeground: 'white',
+    brandBackground: 'red',
+
+    defaultBackground: 'white',
+    defaultForeground: 'black',
+  }
+}
+
+export default globalTokens
+
+// themes/fluent/buttonTokens.ts
+
+const buttonTokens = globalTokens => ({
```

Comment:
_I_ don't like this as it creates a hard coupling between non-treeshakable object `globalTokens` and `buttonTokens`, i.e. it's impossible for the bundler to understand that only `defaultBackground`, `defaultForeground`, `brandBackground` and `brandForeground` required there.

Reply:
What's the alternative proposal? We can import them directly in the `buttonTokens.ts`, but we want to update the tokens if they are updated in the the parent theme context. If we don't have the global tokens, than if we want to update some commonly used thing, we will need to update the tokens to multiple components...

Reply:
If you remember https://codesandbox.io/s/t-rex-proto-1w8bp I had there a proposal to skip merging of tokens & styles. I **don't want** to say that we should go with it, I **want to say** that non-treeshakable tokens are not okay as they will create bloat for any styles:
```
buttonStyles() => (require) buttonTokens() => (require) globalTokens() 
```

Just to get current `siteVariables` I need 5kb of minified code:

![image](https://user-images.githubusercontent.com/14183168/73755881-020f9280-4767-11ea-99d7-a24986747ec2.png)

https://cdn.jsdelivr.net/combine/npm/@fluentui/react@0.43.1/dist/es/themes/teams/siteVariables.min.js,npm/@fluentui/react@0.43.1/dist/es/themes/teams/categoryColors.min.js,npm/@fluentui/react@0.43.1/dist/es/themes/teams/colors.min.js,npm/@fluentui/react@0.43.1/dist/es/utils/fontSizeUtility.min.js

***

However, with direct usages it can be better:

```js
// main.js
import {menuForeground } from './menu'

export const styles = () => ({
  color: menuForeground
})

// globals.js
export const colorBrandForeground = 'white'
export const colorBrandBackground = 'red'

// menu.js
import {colorBrandForeground, colorBrandBackground} from './globals'

export const menuForeground = colorBrandBackground
export const menuBackground = colorBrandForeground
```

Will be minified to:

```js
export default () => ({
    color: "red"
});
```

Demos:
https://rollupjs.org/repl/?version=1.31.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3Qm1lbnVGb3JlZ3JvdW5kJTIwJTdEJTIwZnJvbSUyMCcuJTJGbWVudSclNUNuJTVDbmV4cG9ydCUyMGNvbnN0JTIwc3R5bGVzJTIwJTNEJTIwKCklMjAlM0QlM0UlMjAoJTdCJTVDbiUyMCUyMGNvbG9yJTNBJTIwbWVudUZvcmVncm91bmQlNUNuJTdEKSU1Q24lNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyZ2xvYmFscy5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBjb25zdCUyMGNvbG9yQnJhbmRGb3JlZ3JvdW5kJTIwJTNEJTIwJ3doaXRlJyU1Q25leHBvcnQlMjBjb25zdCUyMGNvbG9yQnJhbmRCYWNrZ3JvdW5kJTIwJTNEJTIwJ3JlZCclMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIybWVudS5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAlN0Jjb2xvckJyYW5kRm9yZWdyb3VuZCUyQyUyMGNvbG9yQnJhbmRCYWNrZ3JvdW5kJTdEJTIwZnJvbSUyMCcuJTJGZ2xvYmFscyclNUNuJTVDbmV4cG9ydCUyMGNvbnN0JTIwbWVudUZvcmVncm91bmQlMjAlM0QlMjBjb2xvckJyYW5kQmFja2dyb3VuZCU1Q25leHBvcnQlMjBjb25zdCUyMG1lbnVCYWNrZ3JvdW5kJTIwJTNEJTIwY29sb3JCcmFuZEZvcmVncm91bmQlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXNtJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=
https://codesandbox.io/s/terser-el1eu

Reply:
Thanks a lot for sharing this @layershifter I remember this prototype and discussions, and from what I remember the scenario that this doesn't support is on the fly changes which needsto be propagated to the children themes. @levithomason  is this still a valid requirement?

The other thing is, theme composition will be a bit hard for managing if the tokens are not part of the theme. Once they become local imports, it will be hard to change anything or keep changes consistently across different themes. I am open for different proposals!

Reply:
After offline discussions, as we still don't have answer for how the things would be propagated down to the children theme if they are not in the context, we will leave this section as is, and in the future if we come up with better strategy, we will update it.